src/lib/device/virtual_dev.cpp

```diff
@@ -502,6 +520,19 @@ void VirtualDev::update_vdev_private(const sisl::blob& private_data) {
     hs_utils::iobuf_free(buf, sisl::buftag::superblk);
 }
 
+///////////////////////// VirtualDev Checkpoint methods /////////////////////////////
+
+VDevCPContext::VDevCPContext(cp_id_t cp_id) : CPContext(cp_id) {}
+
+std::unique_ptr< CPContext > VirtualDev::create_cp_context(cp_id_t cp_id) {
+    return std::make_unique< VDevCPContext >(cp_id);
+}
+
+void VirtualDev::cp_flush(CP*) {
+    // pass down cp so that underlying componnents can get their customized CP context if needed;
+    m_chunk_selector->foreach_chunks([this](cshared< Chunk >& chunk) { chunk->cp_flush(); });
+}
```

Comment:
I think it should use m_all_chunks which will contain all chunks of this virtualdev.
if chunkselector is HeapChunkSelector, once a chunk is selected, it will be removed and only contains part of chunks in this virtualdev.

Reply:
I thought about it, it depends on how chunkselector is maintaining its chunks. E.g. should block_allocator bypass chunkselector to assume that all chunks should participant checkpionting or always let chunkselector to decide. 
Your point is valid that when a chunk is selected, it is removed from the heap, but one design option is heapchunkselector can still have a view of all the chunks that it maintains. Or we can just remove it from chunkselector and let vdev to add it later, which is also a fine design.

Reply:
I see there are 2 options broadly.
Option 1: m_all_chunks is maintained in vdev and for CP and other instrumentation, we iterate over m_all_chunks. chunk_selector exposes a method foreach_active_chunk() which = m_all_chunks for round robin selector, but only unallocated chunks for heap selector
This m_all_chunks needs to be thread safe (for vdev supporting dynamic chunk addition), which will do rcu based
Pros: The vdev doesn't need a lamba or additional things and can iterate and functionality sits in vdev.
Cons: A split responsibility between vdev and chunk selector
(OR)
Option 2: vdev does not maintain any list of chunks, but put the responsibility to chunk_selector. chunkselector base class provides rcu based protection (only for dynamic chunk addition cases). It will expose 2 methods foreach_chunk() and foreach_active_chunk()
Pros: Clean interface and everything is well clear and vdev will be handling off of to chunkselector
Cons: We need a lambda to access blk_allocation which is a very very critical code path. This problem is mitigated if we don't use std::function, but just lambda (which compiler optimizes)

Reply:
yes, I am ok with option 2 too which will make interface more clear and more make sense.   if we are going to do that, we should at least add a function description to foreach_chunk()/foreach_active_chunk() to make the derived class user to know its purpose and difference.

Reply:
@hkadayam  would you please explain more on `Cons: We need a lambda to access blk_allocation which is a very very critical code path. ` ?

Reply:
@hkadayam IMO, the only difference between active chunk and inactive chunk is active chunk will receive allocblk request and inactive chunk not. but they both might have dirty offset, so there is not difference between them for taking checkpoint and i am not sure why we need `foreach_active_chunk()` here

Reply:
@xiaoxichen I think the only thing is std::function has a runtime performance cost which foreach_chunk is taking as input parameter, but can be mitigated by lambda.

Reply:
@JacksonYao287 You are right that `foreach_active_chunk()` is needed only in case we need to do something specific to active chunks and that there are no use case right now. The checkpoint will have a CPContext which will have list of dirtied chunks and persist them irrespective of whether they are active or not.

@xiaoxichen My point was std::function does dynamic memory allocation and make sense only for storing the function pointers in an async call. So we can use plain lambda with (auto) instead of std::function and then the cost is just another function call.

Reply:
Yes but I am not very understanding why this is in the critical code path?   For IO we will not go through any of the for-each right?  ChunkSelector select a chunk and block allocation is made on that chunk only

Reply:
Yes I take that statement back that `alloc_blks()need to iterate the list of chunks`. It does iterate when the chunk selector returns a chunk which doesn't have much space or allocation failed, which is a rarer scenario.

My only concern is that we don't want to walk through all chunks on block allocation time. We do that for get_available_blks() and get_used_blks() which I am hoping is done rarer.

Reply:
We have nblks passed in to chunk selector
`virtual Chunk* select(blk_count_t nblks, const blk_alloc_hints& hints)` which should gareentee/best effort?  the returned chunk has sufficient blks, or , at least the chunk with most allocable blks if none of the chunk has `avail() > nblks`
so yes we should very rare(maybe never) use `foreach_chunks` in allocation path.

Reply:
When allocation fails on a particular shard, HO should seal that shard and create another shard which goes to chunkselector and picks the top chunk in heap. If later blk allocation also fails, it repeat this step until we have picked all the chunks and emptied the heap, in this rare case, I would say GC would be kicked in sometime ealier by some watermark/watchdog. 

in either cases, I don't see a need to call foreach_chunk in either blk allocator or shard creation.

Reply:
Yes we dont need to call foreach for sure.
Through a few nit on the previous statement:
>When allocation fails on a particular shard, HO should seal that shard and create another shard which goes to chunkselector and picks the top chunk in heap. 
HO will report IO error and DM decide to seal,  or HO seal and notify DM . In both case HO should not be the one to create new shard as the `putBlob(shardID,  blobData)` has shardID in the signature. 


Anyway this discussion is clear enough,  as far as it is not in the critical path,  do we see we need to change it to function ptr (lambda) instead of std::function?